### ISSUE TITLE
Teleportation, Landing points, logging into last location

### ISSUE THREAD

**Issue by emperorstarfinder** [HalcyonGrid/halcyon#61]
As this is a persisting issue still, per Ricky's request reposting this here.  As it is late and it would take a while I am not going to copy and paste all the comments from the old issue here:  So please see https://github.com/IslandzVW/halcyon/issues/403 for history.

So while doing some testing I seem to have stumbled on a couple of rather strange (or perhaps not so strange) issues.

Teleportation using world map - When doing a teleport via world map in the viewer (for this purpose we shall say from Parcel 1 located in Region A to Parcel 2 located in Region B, when a landing point is set in Parcel B it appears the function of teleporting using the world map is not respecting a landing point that is set in the parcel details. In testing other architectures which at one time also had OpenSim code (I know halcyon went away from Opensim in 2010 so we don't go down that rabbit hole here), it appears this problem carried over to them as well. Perhaps this is one of those known carry-overs that wasn't known?

When logging out (any viewer appears to produce this) from say a platform at z: 1500 above land and then logging back in the last position is incorrect. In my tests I actually logged back in to being underwater underneath the dock on Region A. The platform I logged out from is in Region B. This also appears to cause trouble with viewers logging in as it complained it couldn't log in and then when I did attempt number 2 to log in it actually did log in. Perhaps a bug of some kind?

My team is working on this issue on our test-bed Halcyon grid.  Per discussion with Ricky in Discord if we find the solution we will pass it on.  Per the comment from @appurist on the previous issue, I will be taking a look at the Avatar Location Protocol as part of my in-depth code review for this issue as it does seem possible that the protocol is not handling some of the updates correctly: i.e. "There's a fairly significant problem with the avatar location protocol where the update to the central grid services does not include the region, just the x/y/z position."

My team has ruled out the possibility of this being a viewer issue as it was reproduced on multiple different viewers.

**Comment by appurist** [comment]
As I discussed privately, but wanted to note here as well, I don't think this has anything to do with the landing point code, but rather this is just a matter of whether the server has received a full region/x/y/z position or not.  If you end up at terrain height, that implies it received zeros, or 128,128,0 which is sometimes what you get if you search for a region name and don't click on the map.  It's likely really an issue with the viewer communicating the location to the region, when the landing point code is **not** enforced. When the landing point code **is** enforced, that would "auto-correct" an incorrect or partial or default x, y, or z.  This would explain why it seems to not be a problem when the landing point is enforced, because the region server handles that. It's only a problem when the location comes from the viewer.

**Comment by emperorstarfinder** [comment]
Well except that argument relies on the fact that if an LP is actually defined that the enforcement is happening.  which it appears not to be happening even if the user is teleporting in from a different region and a landing point has been defined on the destination parcel.  Also if you end up at 128, 128, 0 you actually would land under water (0 on the z axis is the default terrain base of the region not the height).  If you were landing correctly you would land at something like 128, 128, 21 which is about the default height of the terrain just above water without the water bleeding through the terrain texture and is the exact middle of the region (128, 128, 0 being the exact default center of an entire region.

So I can see a situation where the LP is either not being enforced when set or the enforcement information is not being sent to the viewer just like the viewer for some reason isn't sending the full region destination position.  It could even be a combination of the two.  May need to hook up grid proxy to halcyon locally and log in the viewer through the grid proxy software and see what gets returned packet wise to see if ithat sheds some light as well.

**Comment by appurist** [comment]
@emperorstarfinder That all depends. Does the destination have a single parcel defined, or multiple parcels?

I don't follow what you mean about water (water is not relevant here). The only thing that matters is the terrain height where the avatar ends up. If you're standing on the raw terrain, e.g. terrain height+1m, then that means the landing point was not used (unless that's how it was set).  But remember, landing points are per-parcel settings, not per-region, so teleporting within a region to a different parcel is the same as teleporting from another region, in terms of the landing point semantics.  The difference is the viewer to region communications is more complex.  If teleporting within a single parcel (e.g. if it is region-wide), the landing point is not used.

If there's a reproducible case anywhere here, setting a breakpoint (or adding an output log message) in the receiving end will answer what data the destination server is receiving. We could speculate for months, but if there's a reproducible case here, even specific to a region, it's easy enough to start dividing the problem into smaller problems and narrow down the cause.

**Comment by emperorstarfinder** [comment]
> If teleporting within a single parcel (e.g. if it is region-wide), the landing point is not used.

Bingo we have a winner.  This is what this matter is all about.  it should not matter if there are multiple parcels in a region or not.  Parcels are a chunk of land in the region.  If there is only one parcel of land in the entire region it is likely the entire 65536 sqm of land.  Therefore the landing point should be getting recognized.  It should not matter one way or the other if you are teleporting from another region to a parcel in the destination region or from one parcel to another parcel in the same region.  The landing point set via parcel settings should always be recognized and given to the viewer.  The viewer expects the same when it comes to one single region-wide parcel of land in relation to the landing point.  

This is why if you set a landing point in the parcel settings for a parcel in Region A and then teleport to that parcel from Region B you land at the wrong place.  When it comes to what settings a parcel has the viewer does not know about that information until the agent arrives in the parcel and that information is passed to the viewer.  Likewise, during the teleportation phase, the viewer should be given information about the landing point.  

> I don't follow what you mean about water (water is not relevant here).

Look around the outside of your region what do you see? a texture that looks like water moving in one direction right? now, what do you think flows through your terrain? that same water texture.  if you terraform the terrain and dig a hole you will see that water moving through the region.  Now if you have terraformed the terrain to make yourself a lake or shoreline and you place a dock (or just a plain prim for all purposes of what I explained earlier) you will be at a position height of 21 or 22 (position axis z).  Now let us say you've terraformed the terrain to go all the way down to the base of the terrain (position axis z is 0 for the terrain base).  Now this region this parcel is in will be region A.  (you can leave just one parcel in the region).  now if you teleport from a different region (Region B) to the landing point you placed on that object (mesh or non-mesh appear to have the same effect here), you are likely to land not on the object but somewhere under the water (likely at the level you lowered that part of the terrain too).  

Now the viewer does not care one way or the other about how many parcels are in a region during a teleport.  The viewer only cares about the information relating to the destination you are teleporting to.  If a landing point has been set in the parcel (regardless of whether it is one region-wide parcel or smaller parcels) then the viewer will want to know that no matter if you are teleporting from a landmark or place search or world map.  The only way that the landing point should be overridden if your teleporting from any location in Region B to the parcel in Region A is if the teleport was initiated from a teleport offer from another agent (avatar).  Keep in mind we are talking about a normal user whose godLevel flags are 0 and is not the parcel or estate owner or estate manager or a grid god.  

Now what I have described should also be occurring for teleporting between two parcels in the same region.

**Comment by appurist** [comment]
> it should not matter if there are multiple parcels in a region or not. 
It doesn't. I think you misunderstood what I have been saying.  It doesn't matter if there are multiple parcels.  The software doesn't care if there's one parcel or more than one.

However, a Landing Point is a per-parcel option, nothing to do with regions.  I asked if it was a single-parcel region because I was trying to determine if there was any chance at all of a parcel boundary crossing involved, or if I could rule that out.

What _does_ matter for a Landing Point is whether the avatar is _entering a parcel_. The Landing Point location is only enforced when _entering_ a parcel.  (There is no "landing" otherwise.)  As I said above, the first entry into a new (different) parcel than the one you were in previously enforces the landing point for that new parcel. If you're teleporting _within_ a given parcel, there is no parcel entry and the LP is not enforced. This allows teleports within parcels to work.

This was a usability tweak (years ago) that came from many user requests, complaints rather, that teleports were completely broken if there was a landing point defined. Because they were enforced, even when you had already entered the region, so a normal teleport was no longer possible at all. This led to people having to create multiple parcels, just to allow teleports between two points.

Terrain level and water level (which can be anything, not just 20) is not part of landing points at all. However, it is related to teleports, since a minimum of the terrain height is enforced on all teleports.  If you are finding that you're ending up at terrain height (whether above or under water level is not a factor), then that implies that a value of that terrain height or lower (or zero) was specified by the viewer, or stored in a landmark or whatever.  That wouldn't matter in cases where landing points are enforced, since whatever the viewer provides is only used to determine which parcel is the destination.

I think everything above boils down to the last sentence in your first paragraph: a teleport from a different region is likely to end up at terrain height.  That is not my experience with landing points.  However, that **is** the behavior of _not_ having a landing point defined, and the viewer passing a zero for Z or something bogus.

So let's focus on that specific part of the report: _"if you teleport from a different region (Region B) to the landing point you placed on that object (mesh or non-mesh appear to have the same effect here)"_ What do objects have to do with this? Landing points are specified on land parcels, not objects. 

Can you define what you mean by "the landing point you placed on that object"?  It sounds like you aren't even talking about a Landing Point but rather something else.

"The only way that the landing point should be overridden if your teleporting from any location in Region B to the parcel in Region A is if the teleport was initiated from a teleport offer from another agent (avatar)."
That is mostly correct, but in addition to a "teleport lure", it is also true that gods, estate owners, partners and the destination parcel owner don't have Landing Points enforced for that parcel.  If you're testing with one of those, you could end up where ever the viewer requested.

Also, note that in many viewers, the Z position is (and sometimes X and Y position are) not included (e.g. if you search for a region name on the map and then hit Teleport).  That doesn't matter if there's a Landing Point in the parcel containing 128,128, but if there isn't or it's being requested by someone with land authority in the destination parcel, they'd end up at something like 128,128,0 or whatever the terrain height is.  Including under water. 

This is all by design, although there are things I'd like to see improved at the viewer end (such as that use of 128,128,0 as a destination).

**Comment by kf6kjg** [comment]
I understood "the landing point you placed on that object" to mean "the landing point specified in the parcel at my avatar's position when I was standing upon a specific object that may be a small platform well above terrain height."

**Comment by appurist** [comment]
I'm just going to resummarize the key points of Landing Point handling:
1. Landing Points **are** applied/enforced only when _entering_ a parcel, not for teleports within the same parcel.
2. Landing Points **are not** enforced for land authorities: gods, estate owners & partners, estate managers and the destination parcel owner.

Is his issue report requesting that one of those (or both) change?

**Comment by kf6kjg** [comment]
Ok Vin and I just did some quick qualitative testing and each test came up as expected with two possible exceptions that I need to research and do quantitative testing on.

Setup was to have a target region with a single parcel and a secondary region to come from. We also had two accounts: the first account, V, was the admin account and was sued solely to set the parcel landing point settings; the second account, T, was a basic nearly new avatar with no special privileges, permissions, ownerships, groups, etc.  The parcel landing location was set in the target region by placing a small platform at <117,120,30>, about 10m above the terrain height at that location. The teleport routing was set as specified below.  A recent version of Firestorm was used.

Using the T avatar we tried logging in, moving into the region to an arbitrary location and then relogged using "previous location" login screen selection.  Avatar T was found to be at the previous location as expected.  Target parcel's teleport routing setting could be set to either "Anywhere" or "Forced"with no change in behavior.  This seemed to be expected.

Using the T avatar we TP'd to a second region then relogged, but this time specifying the target region's name.  Avatar T was found to be at <128,128, terrainHeight>.  Target parcel's teleport routing setting could be set to either "Anywhere" or "Forced"with no change in behavior.  This was initially unexpected, but made sense as this is a "login" not a "teleport".  Anecdotal evidence indicates SL has the same behavior, but we have not yet verified.

Using the T avatar we TP'd to a second region, then had avatar V set the teleport routing to "Anywhere", and none of our TP tests into the target region's parcel ever landed anywhere near the landing location: scripted TP went where it was told to send the avatar, TP'ing via the map automatically sets a destination of <128,128,2> so we landed at <128,128,terrainHeight> as expected.  In all it seemed to me like setting teleport routing to "Anywhere" makes the landing location largely ignored - there might be cases in this set up that I've not tested where it is used.

Using the T avatar we TP'd to a second region, then had avatar V set the teleport routing to "Forced", and ALL of our quick and in-exhaustive TP tests into the target region's parcel landed right at the landing location: scripted TP, map TP, etc.  This also matches expectations.

As to logging back in on a platform in the sky using "last location" on the login, I have done this regularly enough to know for certain that I always land on my platform - even when my platform is a super thin mesh sitting at 4000m and independent of whether my avatar has any privs or no.  If that was a bug I'd have run across it in any of the Halcyon grids I visit, and I've used both various versions of Firestorm and Alchemy doing this.

Quantitative testing would need to be done and compared to SL's results and Halcyon's design decisions, but thus far I'm not seeing anything that stands out as a bug. Please correct me if I'm wrong and give a detailed reproduction case that specifies the relevant parcel settings, in-world test conditions, and account statuses.

**Comment by appurist** [comment]
@kf6kjg Everything there seems as I would expect. Specifying a region name (only) on a login form of course results in a request from the viewer for the default location and arrival at terrain height as expected.

Some of the tests you did will differ from SL behavior, as per the two points I made above being the result of many "bug" reports and complaints that resulted in a couple of tweaks over the years that people either don't notice or eventually come to like.  (The complaints stopped after the adjustments.)

**Comment by emperorstarfinder** [comment]
I would expect you would be landing on your platform in last login.  I never mentioned last login as I knew that was working properly.  

I did a little test matching this with the current Halcyon, OpenSim, and Secondlife using the latest Viewers (Alchemy, Firestorm 6.0.2. Singularity).  For the record I haven't tested this yet against Kokua Viewer though I should.  The results I get which I have mentioned extensively in Halcyon are reproducible in OpenSim so I am not sure that it is necessarily just the fault of one architecture.  The only place that the LP logic appears to be working as it was intended is in Secondlife.  

Viewers used in this test:

- Firestorm: Versions 5.1.7.55786 through the latest 6.0.2.56680
- Alchemy: Version 5.0.7.41341 (Note because Alchemy has not had a release since November 17, 2017 if you only test against this viewer when developing your architecture you likely will miss things)
- Singularity: Version is the latest nightly build.

For both the latest Halcyon, OpenSim, and Secondlife  The setup for this test is as follows:

NOTE: For all tests the avatar was not the region owner, estate manager, parcel owner, part of a group, or grid god.  The avatar was just a normal avatar.

Halcyon Tests

Test 1:

created two brand new regions (both connected to the grid services),  

Region A had the parcel landing point set at <137, 129, 22>.  The parcel was not deeded nor set to group.  Therefore the new group role ability "ignore landing point" that is now in Firestorm 6.0.2 would not be a show stopper here.  No Terraforming was done.

Region B had nothing special and was used to teleport to Region A.

Teleport from World Map: Double click anywhere on the region tile in the world map and that is where you will land in Region A regardless of the LP setting. (similar to Secondlife though this likely is because the world map cannot necessarily know about parcel settings).

Teleport from Landmark: Landmark appears to land as intended.

Creation of new Avatar and logging in: Avatar lands at <128, 128, 22>  A brand new avatar would not have a last known location so at that point the last known location wouldn't even be a question either.


Test 2: 

created two brand new regions (both connected to the grid services),  

Region A had the parcel landing point set at <137, 129, 22>.  The parcel was not deeded nor set to group.  Therefore the new group role ability "ignore landing point" that is now in Firestorm 6.0.2 would not be a show stopper here.  The terrain around the landing point was terraformed to a height of 0 (terrain base under the water which if you look around the outside of the region you will see the surface of moving through the region.).  The terraforming was done to include the position <128, 128, 0> (which is also the known dead center of a region) to be sure there wasn't something ominous going on that I was missing. A plain square prim was set up with a size of <10, 10, 0.25> at the landing point.  .The object was not mesh nor sculpted.

Region B had nothing special and was used to teleport to Region A.

Teleport from World Map: Double click anywhere on the region tile in the world map and that is where you will land in Region A regardless of the LP setting. (similar to Secondlife though this likely is because the world map cannot necessarily know about parcel settings).

Teleport from Landmark: I landed at <137, 129, 0> which was under the object. and was at the base of the terrain in the region.  

Creation of new Avatar and logging in: Avatar lands at <128, 128, 0>  A brand new avatar would not have a last known location so at that point the last known location wouldn't even be a question either.


OpenSim Tests

NOTE: Testing of this was done in OpenSim for comparison to Halcyon and Secondlife behavior as I remembered this issue occurred in OpenSim 0.6.5 and appears to still be an issue in OpenSim 0.9.1 It should be noted Halcyon forked away with OpenSim 0.6.5 as referenced by the AssemblyInfo.cs files in the OpenSim directory of the Halcyon source code.

NOTE 2: Both tests performed in Halcyon were repeated in OpenSim using Alchemy, Firestorm, and Singularty viewers with the same exact versions used with the Halcyon tests.


Test 1:

created two brand new regions (both connected to the grid services),  

Region A had the parcel landing point set at <137, 129, 22>.  The parcel was not deeded nor set to group.  Therefore the new group role ability "ignore landing point" that is now in Firestorm 6.0.2 would not be a show stopper here.  No Terraforming was done.

Region B had nothing special and was used to teleport to Region A.

Teleport from World Map: Double click anywhere on the region tile in the world map and that is where you will land in Region A regardless of the LP setting. (similar to Secondlife though this likely is because the world map cannot necessarily know about parcel settings).

Teleport from Landmark: Landmark appears to land as intended.

Creation of new Avatar and logging in: Avatar lands at <128, 128, 22>  A brand new avatar would not have a last known location so at that point the last known location wouldn't even be a question either.


Test 2: 

created two brand new regions (both connected to the grid services),  

Region A had the parcel landing point set at <137, 129, 22>.  The parcel was not deeded nor set to group.  Therefore the new group role ability "ignore landing point" that is now in Firestorm 6.0.2 would not be a show stopper here.  The terrain around the landing point was terraformed to a height of 0 (terrain base under the water which if you look around the outside of the region you will see the surface of moving through the region.).  The terraforming was done to include the position <128, 128, 0> (which is also the known dead center of a region) to be sure there wasn't something ominous going on that I was missing. A plain square prim was set up with a size of <10, 10, 0.25> at the landing point.  .The object was not mesh nor sculpted.

Region B had nothing special and was used to teleport to Region A.

Teleport from World Map: Double click anywhere on the region tile in the world map and that is where you will land in Region A regardless of the LP setting. (similar to Secondlife though this likely is because the world map cannot necessarily know about parcel settings).

Teleport from Landmark: I landed at <137, 129, 0> which was under the object. and was at the base of the terrain in the region.  

Creation of new Avatar and logging in: Avatar lands at <128, 128, 0>  A brand new avatar would not have a last known location so at that point the last known location wouldn't even be a question either.


Secondlife Tests

For these tests it should be noted I used regions a member of my family owns so we can thank them for allowing me to test this without complaint.

Test 1:

In Region A we set a similar landing Point at <137, 129, 22> in the parcel settings.  As before the new group role ability "ignore landing point" was not at issue here (though probably should be tested at some point as both Singularity and Alchemy currently do not have this feature as it is new).  The first test there was no terraforming done.

In Region B no changes were made as it was the region being teleported from.

Results:

World Map: Teleporting from the world map you can double click on any position in the region and teleport to it.  This appears to be normal behavior for Secondlife as its obvious that the world map would not necessarily know about landing points.

Teleporting from landmark: When teleporting to Region A via landmark the landing was dead on the set landing point in the parcel settings.  I also created landmarks that were not at the landing point to see if the LP was overriding the landmark and this landmark also went straight to the landing point.  Therefore Secondlife does override the place landmarks are created if its different from a set landing point as it should be.

Created new Avatar: Unfortunately creating a new avatar in Secondlife would land you at one of their welcome islands so that test couldn't be undertaken without a little hassle.


Test 2:

In Region A I proceeded to terraform as I had done before with the OpenSim and Halcyon tests.  I created a normal square object of size <10, 10, 0.25> as before and put it at the landing point of <137, 129, 22> and set the landing point.  (terrain was restored to the way it was after all testing complete to be respectful of family member).

In Region B there were no changes made as it was only a region being used to teleport from.

Results:

World Map: Double clicking on any position on the world map in Region A landed you at that position as expected.  It is very much likely that the world map does not necessarily nor can it necessarily know about a landing point being set at the parcel settings level.

Teleport from Landmark: As with test 1, when teleporting to Region A via landmark the landing was dead on the set landing point in the parcel settings.  I also created landmarks that were not at the landing point to see if the LP was overriding the landmark and this landmark also went straight to the landing point.  Therefore Secondlife does override the place landmarks are created if its different from a set landing point as it should be.

Created new Avatar: Unfortunately creating a new avatar in Secondlife would land you at one of their welcome islands so that test couldn't be undertaken without a little hassle.

The next step I will take here is to look at the teleportation logic in the viewer source codes to see what might be different there and compare that to the Halcyon source code as well as OpenSim.  It is quite possible that even with the fixes that Jim mentioned InWorldz did that it is not enough with all the recent changes in the protocols in the viewers that this issue probably was resolved and is probably returned in the latest viewers.  

While I do believe Rick when he says he tried teleporting to his platform, I cannot ignore the test results I see with my own eyes.  Likewise, I am sure InWorldz probably did make changes in their teleportation logic from the time they forked away from OpenSim to the time the code was placed on Github.  Unfortunately changes during that time are not publicly available, so I cannot realistically review those changes myself and judge as to whether they were right or wrong.  We also have to take into account that with time things change just as the coding logic.  It is possible (though I don't remember paying attention to the LP issues in InWorldz back in the day), that it actually did work similar to Secondlife, but with recent updates to viewer protocols it could even be that it got broken again without anyone knowing.  I also know from experience that while one (or more people) can produce similar results, others will not be able to.  There are always issues that may exist but just cannot be reproduced by everyone and will not necessarily throw you messages in consoles to say "hey this is an error you seriously need to fix me".  

The original intent of Landing Points was to give parcel owners a means of controlling where visitors land on their parcels.  This is because otherwise visitors could land in someone's home unannounced or land in a location on a parcel the parcel owner may otherwise not want them to land at.

WIll report fruther as I continue looking into this matter.

**Comment by kf6kjg** [comment]
Good tests EmperorStarfinder, however I see one critical omitted detail: the value of the teleport routing setting in each test.  Without the value of that being set the same in each test condition the results can vary wildly. Superficially it looks like in your Halcyon and OpenSim tests it was set to Anywhere and in SecondLife it was set to Forced.

**Comment by emperorstarfinder** [comment]
It was omitted as it was set the same way in all 3.  The only thing I did forget is to test it both ways (forced and set to anywhere).  Ill have to test it that way to and see if maybe this comes down to a setting that should be sorted by default.

**Comment by emperorstarfinder** [comment]
Okay this appears to be the routing setting that is in fact the culprit.  If the routing point is set to Landing Point then this does in fact work.  However might consider making this something that gets set automatically when parcel owners set a landing point.  This way the problem does not get run into.  But I will have to look at this further when there is time.  Its not a show stopper so would only really be a minor enhancement.  It also might be something I will have to consider in the new viewer being written by my team.

**Comment by kf6kjg** [comment]
That is expected behavior and in that specific manner, for what it's worth, it matches SL: Teleport Routing set to Anywhere means "there is no landing point" even if it is set.  Consider it a way to turn off the LP without losing it.  A setting of "Blocked" means "no teleports into this parcel".

That said, there are significant differences between SL and IW's details on what exactly happens, but both thus far seem logically sound in their own way.

Shall I mark this as working as intended then?

**Comment by appurist** [comment]
As @kf6kjg points out, the landing point and the routing setting are two different things, although related.  The location is just the location that would be used **_if_** a landing point was enabled.

If I recall correctly, the LP is not _completely_ unused if one is set yet routing is not set to "Landing Point": I think it is used as the _default_ location for parcel-related operations where a specific location is not provided. e.g. If I recall correctly, this is the case for adding a Classified ad on a user profile.  

I would consider this working as designed (closable).

**Comment by emperorstarfinder** [comment]
Well the current code is working, though it probably can use some refinement.  I would prefer to see this work where when the parcel owner sets the landing point, the default routing is changed automatically to reflect that people should be landing at the landing point.

And to clarify from my testing so everyone understands.  The level of a user account should not matter (i.e. GodLevel, Parcel Owner, Region Owner, Estate Manager, etc.).  In testing the proper course is that if a landing point is set and the account has a god level set in their viewer then and only then should a parcel's landing point be overridden.  The intention from LL for that was to help ensure parcel owner's have control over privacy and other such things (Roleplay lifestyles often want you to land in a safe zone, Stores like you landing at the customer service desk or outside the door so you have to walk in and walk around more, etc.).

So as I said, it would be a minor enhancement to have the routing setting change automatically if a landing point is set.  This would not necessarily be a show stopper thing.  On my My-Halcyon repo I changed the corresponding issue to this to Enhancement to remind me to look at later, as I would rather look at the more pressing and major things that have to be done first and the current logic works for the time being.

**Comment by kf6kjg** [comment]
To clean things up could you open that request in a new issue?  That way all the above confusion will go away.

I also intend to finish off my own documentation pass on TP routing and publish it on the wiki here.
